### ISSUE TITLE
CORS Issue AsyncHttpConsumer

### ISSUE THREAD

**Issue by waqasraz** [django/channels#1247]
```Access to XMLHttpRequest at 'http://127.0.0.1:8001/test_middleware/' from origin 'http://127.0.0.1:8082' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.```

Adding a header to each send response solve the above problem but I don't want to add a header to each response. 

```
self.send_response(200, b"xxx"
headers=[
                (b"Content-Type", b"application/json"),
                (b"Access-Control-Allow-Origin", b"*"),
            ]
)
```

**Comment by Zarathustra2** [comment]
Seems like you want this: [Django-Cors-Headers](https://github.com/ottoyiu/django-cors-headers)

**Comment by carltongibson** [comment]
Thanks @Zarathustra2, yes.

**Comment by matthiask** [comment]
I don't think django-cors-headers will work since, at least in my understanding, Django's middleware does not run when you're using `AsyncHttpConsumer`.

@waqasraz Maybe you have some luck using ASGI middleware from the Starlette project, e.g. https://www.starlette.io/middleware/#cors-preflight-requests

**Comment by Zarathustra2** [comment]
@matthiask totally forget about it, @carltongibson maybe reopen the ticket again and @waqasraz could you provide a minimal project for reproducing that issue?

**Comment by waqasraz** [comment]
@Zarathustra2 
@matthiask is right middleware don't run when sending a request to `AsyncHttpConsumer`.

**Comment by softmarshmallow** [comment]
any updates for this?

**Comment by sidharthramesh** [comment]
All the http requests still seem to go through the `MIDDLEWARE` in settings.py if the `application = ProtocolTypeRouter()` does not define anything for http. I've some websocket consumers, but the http requests seem to go through the django-cors-headers as expected.

**Comment by matthiask** [comment]
@waqasraz Yes that's correct.

@sidharthramesh Requests which are handled by `AsyncHttpConsumer` do not go through Django at all (only through channels) and because of this Django's middleware does not run. So yes, django-cors-headers still handles requests to Django views but the same isn't true for equests handled by custom channels consumers.

**Comment by carltongibson** [comment]
This still seems the right answer: 

> Maybe you have some luck using ASGI middleware from the Starlette project, e.g. https://www.starlette.io/middleware/#cors-preflight-requests

Some ASGI middleware will be needed. This is probably out of scope for channels itself.

**Comment by jray-afk** [comment]
Hi all, do you know how to mount the CORSMiddleware component from Starlette onto a Django Channels app?

At the Starlette documentation linked above (https://www.starlette.io/middleware/#cors-preflight-requests) there is a section at the bottom "Using middleware in other frameworks" where I tried to follow their suggestion, but I'm still getting a CORS error from my React app.

Here is my routing.py file:

```
from django.conf.urls import url
from channels.routing import ProtocolTypeRouter, URLRouter
from channels.auth import AuthMiddlewareStack
from channels.security.websocket import AllowedHostsOriginValidator, OriginValidator
from myapp.consumers import MyConsumer
from starlette.middleware.cors import CORSMiddleware

application = ProtocolTypeRouter({
    'websocket': AllowedHostsOriginValidator(
        AuthMiddlewareStack(
            URLRouter(
                [
                    url(r"my/url/pattern", MyConsumer()),
                ]
            )
        )
    )
})
 
application = CORSMiddleware(application, allow_origins=['127.0.0.1'])
```

I have also tried the same in my asgi.py file, still getting CORS error:
```
import os
import django
from channels.routing import get_default_application
from starlette.middleware.cors import CORSMiddleware

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "cfehome.settings")
django.setup()

application = get_default_application()
application = CORSMiddleware(application, allow_origins=['127.0.0.1'])
```

**Comment by jray-afk** [comment]
Ah whoops, just took a fresh look at this and realized I was missing the CORSMiddleware component from my wsgi.py as well (implemented about the same as how the asgi.py file looks). Everything is working now and getting no CORS error! Going to leave my previous comment up in case someone else runs into the same problem.

**Comment by carltongibson** [comment]
I'm going to convert this to a discussion, since I don't think it's an Issue per se. 

Happy to link to a blog post or package wrapping up what's needed here.
